src/freenet/client/filter/GIFFilter.java

```diff
@@ -30,22 +35,452 @@
 	public void readFilter(InputStream input, OutputStream output, String charset, HashMap<String, String> otherParams,
 	        FilterCallback cb) throws DataFilterException, IOException {
 		DataInputStream dis = new DataInputStream(input);
-		// Check the header
-		byte[] headerCheck = new byte[HEADER_SIZE];
-		dis.readFully(headerCheck);
-		if ((!Arrays.equals(headerCheck, gif87aHeader)) && (!Arrays.equals(headerCheck, gif89aHeader))) {
-			throwHeaderError(l10n("invalidHeaderTitle"), l10n("invalidHeader"));
-		}
-		output.write(headerCheck);
-		FileUtil.copy(dis, output, -1);
+		try {
+			// Check the header
+			byte[] headerCheck = new byte[HEADER_SIZE];
+			dis.readFully(headerCheck);
+			final boolean isGIF87a = Arrays.equals(headerCheck, gif87aHeader);
+			final boolean isGIF89a = Arrays.equals(headerCheck, gif89aHeader);
+			if (!isGIF87a && !isGIF89a) {
+				throwDataError(l10n("invalidHeaderTitle"), l10n("invalidHeader"));
+			}
+			output.write(headerCheck);
+			if (isGIF87a) {
+				GIF87aValidator.filter(input, output);
+			} else if (isGIF89a) {
+				GIF89aValidator.filter(input, output);
+			}
+		} catch (EOFException e) {
+			throwDataError(l10n("unexpectedEOFTitle"), l10n("unexpectedEOF"));
+		}
 		output.flush();
 	}
 
+	private static abstract class GIFValidator {
+		private final InputStream input;
+		private final OutputStream output;
+
+		// Screen descriptor data
+		protected int screenWidth;
+		protected int screenHeight;
+		protected int screenFlags;
+		protected int screenColors; // Parsed from screenFlags
+		protected int screenBackgroundColor;
+		protected int screenAspectRatio;
+
+		// ",": Image separator character
+		private static final int IMAGE_SEPARATOR = 0x2C;
+		// ";": GIF terminator
+		private static final int GIF_TERMINATOR = 0x3B;
+		// "!": GIF Extension Block Introducer
+		protected static final int EXTENSION_INTRODUCER = 0x21;
+
+		protected GIFValidator(InputStream input, OutputStream output) {
+			this.input = input;
+			this.output = output;
+		}
+
+		/** Checks whether the parsed screen descriptor is valid. */
+		protected boolean validateScreenDescriptor() {
+			if (screenBackgroundColor >= screenColors) {
+				// Not in the specification, but here the background color would refer to a color
+				// index not within the bounds of the Global Color Table.
+				return false;
+			}
+			return true;
+		}
+
+		/** Checks whether the given image flags are valid. */
+		protected boolean validateImageFlags(int imageFlags) {
+			return true;
+		}
+
+		/** Filters the next extension blocks, and skips it when it is unsupported or invalid. */
+		protected void filterExtensionBlock() throws IOException {
+			skipExtensionBlock();
+		}
+
+		/** Signals that image data is found. If the image data is valid, it will be written
+		  * immediately after this method returns. */
+		protected void foundImageData(boolean valid) throws IOException {
```

Comment:
Its purpose is primarily to signal that image data (a render block) has been found. Some extension blocks (such as the graphic control) may only appear once before a render block, and only apply to the render block they precede. See `GIF89aFilter.foundImageData` for details: there's stuff it has to do regardless of the image data being valid (i.e. invalidating a cached graphic control extension block, because it should not apply to the *next* render block even if this one is invalid), and stuff that depends on the data being valid.

Given the simplicity of  `GIF89aFilter.foundImageData`, the sole method that overrides this method, I don't think the additional code for having two different methods is worth it, nor do I think that it would substantially improve readability.